core/src/main/java/io/questdb/log/TemplateParser.java

```diff
@@ -0,0 +1,244 @@
+                            addEnvComponent(dollarStart, dollarStart + 1, i);
+                            lastExprEnd = i + 1;
+                        } else {
+                            throw new LogError("Unexpected '$' at position " + i);
+                        }
+                    } else {
+                        if (i - lastExprEnd > 0) {
+                            addStaticComponent(lastExprEnd, i);
+                            lastExprEnd = i + 1;
+                        }
+                    }
+                    dollarStart = i;
+                    break;
+                case '{':
+                    curlyBraces++;
+                    if (dollarStart == NIL) {
+                        continue;
+                    }
+                    keyStart = i + 1;
+                    break;
+                case '}':
+                    curlyBraces--;
+                    if (dollarStart == NIL) {
+                        continue;
+                    }
+                    if (keyStart == NIL) {
+                        addEnvComponent(dollarStart, dollarStart + 1, i);
+                        lastExprEnd = i;
+                    } else {
+                        int exprLen = i - keyStart;
+                        if (exprLen == 0) {
+                            throw new LogError("Missing expression at position " + keyStart);
+                        }
+                        int formatStart = keyStart + DATE_FORMAT_KEY.length();
+                        if (Chars.startsWith(originalTxt, keyStart, formatStart, DATE_FORMAT_KEY)) {
+                            addDateComponent(formatStart, i);
+                        } else {
+                            addEnvComponent(dollarStart, dollarStart + 2, i);
+                        }
+                        lastExprEnd = i + 1;
+                    }
+                    keyStart = NIL;
+                    dollarStart = NIL;
+            }
+        }
+        if (dollarStart == NIL) {
+            if (curlyBraces != 0) {
+                throw new LogError("Mismatched '{}' at position " + lastExprEnd);
+            }
+            if (locationLen - lastExprEnd > 0) {
+                addStaticComponent(lastExprEnd, locationLen);
+            }
+        } else {
+            if (keyStart != NIL) {
+                throw new LogError("Missing '}' at position " + locationLen);
+            }
+            if (locationLen - dollarStart > 1) {
+                addEnvComponent(dollarStart, dollarStart + 1, locationLen);
+            } else {
+                throw new LogError("Unexpected '$' at position " + dollarStart);
+            }
+        }
+        return this;
+    }
+
+    public void setDateValue(long dateValue) {
+        this.dateValue.set(dateValue);
+    }
+
+    public int getKeyOffset(CharSequence key) {
+        return envStartIdxs.get(key); // relative to originalTxt
+    }
+
+    public ObjList<Component> getComponents() {
+        return txtComponents;
```

Comment:
inconsistent naming; perhaps we give more descriptive name for `Component` class itself?